lib/ui/window.dart

```diff
@@ -143,14 +161,234 @@ class Locale {
+       _scriptCode = script,
+       _countryCode = region,
+       _variants = collection.LinkedHashSet<String>.from(variants ?? <String>[]),
+       _extensions = collection.LinkedHashMap<String, String>.from(extensions ?? <String, String>{});
+
+  /// Parses [Unicode Locale
+  /// Identifiers](https://www.unicode.org/reports/tr35/#Identifiers).
+  ///
+  /// This method does not parse all BCP 47 tags. See [BCP 47
+  /// Conformance](https://www.unicode.org/reports/tr35/#BCP_47_Conformance) for
+  /// details.
+  ///
+  /// TODO: we should try to support parsing all BCP 47 tags, even if we produce
+  /// only Unicode BCP47 Locale Identifiers as output.
+  factory Locale.parse(String localeId) {
+    assert(localeId != null);
+    localeId = localeId.toLowerCase();
+    if (localeId == 'root')
+      return Locale._internal('und');
+
+    final List<String> localeSubtags = localeId.split(_reSep);
+    String language, script, region;
+    final collection.SplayTreeSet<String> variants =
+        collection.SplayTreeSet<String>();
+    // Using a SplayTreeMap for its automatic key sorting.
+    final collection.SplayTreeMap<String, String> extensions =
+        collection.SplayTreeMap<String, String>();
+
+    final List<String> problems = <String>[];
+    if (_reLanguage.hasMatch(localeSubtags[0])) {
+      language = _replaceDeprecatedLanguageSubtag(localeSubtags.removeAt(0));
+    } else if (_reScript.hasMatch(localeSubtags[0])) {
+      // Identifiers without language subtags aren't valid BCP 47 tags and
+      // therefore not intended for general interchange, however they do match
+      // the LDML spec.
+      language = 'und';
+    } else {
+      problems.add('"${localeSubtags[0]}" is an invalid language subtag');
+    }
+    if (localeSubtags.isNotEmpty && _reScript.hasMatch(localeSubtags[0])) {
+      script = _capitalize(localeSubtags.removeAt(0));
+    }
+    if (localeSubtags.isNotEmpty && _reRegion.hasMatch(localeSubtags[0])) {
+      region = localeSubtags.removeAt(0).toUpperCase();
+    }
+    while (localeSubtags.isNotEmpty && _reVariant.hasMatch(localeSubtags[0])) {
+      variants.add(localeSubtags.removeAt(0));
+    }
+
+    // Now we should be into extension territory, localeSubtags[0] should be a singleton.
+    if (localeSubtags.isNotEmpty && localeSubtags[0].length > 1) {
+      final List<String> mismatched = <String>[];
+      if (variants.isEmpty) {
+        if (region == null) {
+          if (script == null) {
+            mismatched.add('script');
+          }
+          mismatched.add('region');
+        }
+        mismatched.add('variant');
+      }
+      problems.add('unrecognised subtag "${localeSubtags[0]}": is not a '
+          '${mismatched.join(", ")}');
+    }
+    _parseExtensions(localeSubtags, extensions, problems);
+
+    if (problems.isNotEmpty)
+      throw LocaleParseException('Locale Identifier $localeId is invalid: '
+                                 '${problems.join("; ")}.');
```

Comment:
nit: flutter style (https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo) would have this be in a `{ ... }` block since it spans multiple lines